### PR TITLE
Performance (instanced rendering)

### PR DESCRIPTION
**Pull request by jytang**
- Framerate increased xfold (where x is 10000 minus the number of seconds it takes to get this pr merged) 
- Tiles are no longer children of grid, for instanced rendering purposes
- Tiles are no longer independent graphical entities, only *one of each tile type* actually gets its draw call called every frame.
- Need to fix grid hovering/selecting (especially green highlight thing), because it depended on each Tile being independent graphical entities.

### CODE REVIEW

File: src/aldente_client.cpp [hunk 1]
```diff
@@ -106,7 +106,16 @@ void AldenteClient::start() {
     Timer timer(GAME_TICK);
     Timer::provide(&timer);
 
+    // Setup framerate counter.
+    GLuint frame = 0;
+    Timer::get()->do_every(std::chrono::seconds(1), [&](){
```
Comment (jytang) [inline src/aldente_client.cpp:115]: is this an okay way to do Timer-based fps counter?

**Review by richard1 — COMMENTED** [review]
Excellent work!  I'm averaging around 60 fps.  As you mentioned, I think the only thing broken is the hover block during building, but all other features work as expected.

I'll leave the graphics review to Kavin and Austin.

File: src/model/geometry.cpp [hunk 2]
```diff
@@ -61,6 +63,36 @@ void Geometry::populate_buffers() {
...
+        glVertexAttribPointer(8, 4, GL_FLOAT, GL_FALSE, 4 * sizeof(glm::vec4), (GLvoid*)(3 * sizeof(glm::vec4)));
+
+        // Setup defaults to be identity matrix.
+        glVertexAttrib4f(5, 1.f, 0.f, 0.f, 0.f);
+        glVertexAttrib4f(6, 0.f, 1.f, 0.f, 0.f);
+        glVertexAttrib4f(7, 0.f, 0.f, 1.f, 0.f);
+        glVertexAttrib4f(8, 0.f, 0.f, 0.f, 1.f);
+
+        // Step forward by 1 for every instance
+        glVertexAttribDivisor(5, 1);
```
Comment (jytang) [inline src/model/geometry.cpp:90]: Austin's machine can't handle this tier of gl.

Comment (jytang) [reply src/model/geometry.cpp:90]: Do not merge yet.

File: src/aldente_client.cpp (hunk 1 above)
Comment (metakirby5) [reply src/aldente_client.cpp:115]: cool & good

File: src/debug_input.cpp [hunk 3]
```diff
@@ -113,7 +113,8 @@ void DebugInput::debug_movement_keys() {
 }
 
 void DebugInput::debug_movement_mouse(events::WindowCursorData d) {
-
+	if (keys[GLFW_KEY_LEFT_CONTROL])
```
Comment (metakirby5) [inline src/debug_input.cpp:116]: What exactly is this?

File: src/model/model.cpp [hunk 4]
```diff
@@ -25,6 +25,32 @@ void Model::draw(Shader *shader, SceneInfo &scene_info, glm::mat4 to_world) {
     }
 }
 
+// Draw multiple instances of this model according to the number of (transformation) mat4s in instance_matrix
+void Model::draw_instanced(Shader *shader, SceneInfo &scene_info,
+                    std::vector<glm::mat4> &instance_matrix) {
+    for (Mesh *mesh : meshes) {
+        // Use model's attached shader if a NULL shader is passed in.
+        if (shader != NULL) {
```
Comment (metakirby5) [inline src/model/model.cpp]: Use `nullptr`?

**Review by metakirby5 — COMMENTED** [review]
Works on mac, LGTM
Preferred if you get a review from a graphics guy as well

File: src/debug_input.cpp (hunk 3 above)
Comment (jytang) [reply src/debug_input.cpp:116]: i just needed a way to take a screenshot of the game with Windows Snipping Tool without my mouse movement causing the game's camera to move. lol.

**Review by Pukami — APPROVED** [review]
It doesn't run on my computer due to my graphics card, but it's fine.  Having this will definitely fix almost all performance issues we'd have. I can still work with it.

**Comment by richard1** [comment]
Works on @AznCelebration's machine

**Pull request merged**